### PR TITLE
Martin's week 1 first four exercises

### PR DESCRIPTION
**Pull request by aaaa0441**
Teacher @shineli1984, please help review this WIP. Please let me know where my syntax & logic can improve. Thanks!

### CODE REVIEW

File: cis194/week1/martin/homework/homework.hs [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+-- Exercise 3
+flatten :: [[Integer]] -> [Integer]
+flatten []          = []
+flatten [[]]        = []
+flatten ((x:[]):[]) = [x]
+flatten ((x:ys):[]) = x : flatten [ys]
+flatten (x:ys)      = flatten [x] ++ flatten ys
+
+sumDigits :: [Integer] -> Integer
+sumDigits x = sum . flatten $ map toDigits x
```
Comment (sinogermany) [inline cis194/week1/martin/homework/homework.hs]: Great to see function composition here :)  
We can simplify it by avoiding pattern matching:
```haskell
sumDigits :: [Integer] -> Integer
sumDigits = sum . flatten . map toDigits
```
BTW, try going to https://hoogle.haskell.org and see if you can find something useful with the query 
```haskell
[a] -> (a -> [b]) -> [b]
```

**Comment by sinogermany** [comment]
It's great overall @aaaa0441 , one of the nice things about haskell is that as long as your functions are exhaustive and it compiles, you will rarely, if ever, run into runtime issues. Then you can refactor depending on how much budget the team has ;)

**Comment by aaaa0441** [comment]
Cool, thanks @sinogermany ! I'll update the PR based on your comments!

File: cis194/week1/martin/homework/homework.hs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+toDigits :: Integer -> [Integer]
+toDigits x
+  | x <= 0      = []
+  | x < 10      = [x]
+  | otherwise   = toDigits(x `div` 10) ++ [x `mod` 10]
+
+reverseIntegers :: [Integer] -> [Integer]
+reverseIntegers []       = []
+reverseIntegers (x:[])   = [x]
+reverseIntegers (x:y)    = reverseIntegers(y) ++ [x]
```
[the remark this answers is not in the document]
Comment (aaaa0441) [reply cis194/week1/martin/homework/homework.hs]: Good point! That's much more meaningful naming!

[the remark this answers is not in the document]
Comment (stevemao) [reply cis194/week1/martin/homework/homework.hs]: Yeah, I should use these naming convention too.

File: cis194/week1/martin/homework/homework.hs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+doubleEveryOther x = reverseIntegers (doubleEveryOtherFromLeft (reverseIntegers x))
+
+
+-- Exercise 3
+flatten :: [[Integer]] -> [Integer]
+flatten []          = []
+flatten [[]]        = []
+flatten ((x:[]):[]) = [x]
+flatten ((x:ys):[]) = x : flatten [ys]
+flatten (x:ys)      = flatten [x] ++ flatten ys
```
Comment (shineli1984) [inline cis194/week1/martin/homework/homework.hs]: List is recursive in it’s tail. So we can think of flatten as in taking the first element out, which is a list, then concat(++) that with the flattened tail.

**Pull request closed without merging**